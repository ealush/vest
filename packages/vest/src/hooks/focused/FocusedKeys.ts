export enum FocusModes {
  ONLY,
  SKIP,
}
