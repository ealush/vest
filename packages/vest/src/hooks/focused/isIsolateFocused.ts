import type { Isolate } from 'vestjs-runtime';

import { VestIsolateType } from 'VestIsolateType';
import type { IsolateFocused } from 'focused';

export function isIsolateFocused(isolate: Isolate): isolate is IsolateFocused {
  return isolate.type === VestIsolateType.Focused;
}
