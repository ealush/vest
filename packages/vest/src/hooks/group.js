import context from 'ctx';
import isFunction from 'isFunction';
import isStringValue from 'isStringValue';
import throwError from 'throwError';

/**
 * Runs a group callback.
 * @param {string} groupName
 * @param {Function} tests
 */
const group = (groupName, tests) => {
  if (!isStringValue(groupName)) {
    throwError(
      `group initialization error. Expected "${groupName}" to be a string.`
    );
  }

  if (!isFunction(tests)) {
    throwError(
      `group initialization error. Expected "${tests}" to be a function.`
    );
  }

  // Running with the context applied
  context.bind({ groupName }, tests)();
};

export default group;
