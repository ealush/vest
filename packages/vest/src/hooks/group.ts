import isFunction from 'isFunction';
import isStringValue from 'isStringValue';
import throwError from 'throwError';

import context from 'ctx';

/**
 * Runs a group callback.
 */
export default function group(groupName: string, tests: () => any): void {
  if (!isStringValue(groupName)) {
    throwGroupError(groupName);
  }

  if (!isFunction(tests)) {
    throwGroupError(tests);
  }

  // Running with the context applied
  context.bind({ groupName }, tests)();
}

function throwGroupError(value: unknown) {
  throwError(
    __DEV__
      ? `group initialization error. Expected "${value}" to be a string.`
      : 'group name must be a string'
  );
}
