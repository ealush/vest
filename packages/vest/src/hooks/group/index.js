import context from '../../core/context';
import validateSuiteParams from '../../lib/validateSuiteParams';
import { isGroupExcluded } from '../exclusive';

/**
 * Registers a group in state.
 * @param {string} groupName
 */
const registerGroup = groupName => {
  const { stateRef } = context.use();
  stateRef.patch(state => {
    const nextState = { ...state };
    nextState.groups[groupName] = state.groups[groupName] || {};
    return nextState;
  });
};

/**
 * Runs a group callback.
 * @param {string} groupName
 * @param {Function} tests
 */
const group = (groupName, tests) => {
  // TODO: Replace with something local
  validateSuiteParams('group', groupName, tests);

  const { stateRef } = context.use();
  const state = stateRef.current();

  if (!isGroupExcluded(state, groupName)) {
    registerGroup(groupName);
  }

  context.run(
    {
      groupName,
    },
    () => tests()
  );
};

export default group;
