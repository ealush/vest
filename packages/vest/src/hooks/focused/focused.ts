import {
  asArray,
  Maybe,
  OneOrMoreOf,
  noop,
  Nullable,
  isNotEmpty,
  isStringValue,
} from 'vest-utils';
import { IsolateSelectors, TIsolate, Isolate } from 'vestjs-runtime';

import { FocusModes } from 'FocusedKeys';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { VestIsolateType } from 'VestIsolateType';

export type ExclusionItem = Maybe<OneOrMoreOf<string>>;
export type FieldExclusion<F extends TFieldName> = Maybe<OneOrMoreOf<F>>;
export type GroupExclusion<G extends TGroupName> = Maybe<OneOrMoreOf<G>>;

export type TIsolateFocused = TIsolate & {
  focusMode: FocusModes;
  match: FieldExclusion<TFieldName>;
  matchAll: boolean;
};

export function IsolateFocused(
  focusMode: FocusModes,
  match?: true | FieldExclusion<TFieldName>
): TIsolateFocused {
  return Isolate.create(VestIsolateType.Focused, noop, {
    focusMode,
    match: asArray(match).filter(isStringValue),
    matchAll: match === true,
  });
}

export class FocusSelectors {
  static isSkipFocused(
    focus: Nullable<TIsolateFocused>,
    fieldName?: TFieldName
  ): boolean {
    return (
      focus?.focusMode === FocusModes.SKIP &&
      (hasFocus(focus, fieldName) || focus.matchAll === true)
    );
  }
  static isOnlyFocused(
    focus: Nullable<TIsolateFocused>,
    fieldName?: TFieldName
  ): boolean {
    return focus?.focusMode === FocusModes.ONLY && hasFocus(focus, fieldName);
  }

  static isIsolateFocused(isolate: TIsolate): isolate is TIsolateFocused {
    return IsolateSelectors.isIsolateType(isolate, VestIsolateType.Focused);
  }
}

/**
 * Adds a field or a list of fields into the inclusion list
 *
 * @example
 *
 * only('username');
 */
// @vx-allow use-use
export function only(match: FieldExclusion<TFieldName> | false) {
  return IsolateFocused(FocusModes.ONLY, defaultMatch(match));
}
/**
 * Adds a field or a list of fields into the exclusion list
 *
 * @example
 *
 * skip('username');
 */
// @vx-allow use-use
export function skip(match: FieldExclusion<TFieldName> | boolean) {
  return IsolateFocused(FocusModes.SKIP, defaultMatch(match));
}

function defaultMatch(match: FieldExclusion<TFieldName> | boolean) {
  return match === false ? [] : match;
}

function hasFocus(focus: Nullable<TIsolateFocused>, fieldName?: TFieldName) {
  return (
    isNotEmpty(focus?.match) &&
    (fieldName ? focus?.match?.includes(fieldName) ?? true : true)
  );
}
