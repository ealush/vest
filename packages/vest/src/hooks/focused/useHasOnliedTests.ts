import { isNotNullish } from 'vest-utils';
import { TIsolate, Walker } from 'vestjs-runtime';

import { TIsolateTest } from 'IsolateTest';
import { TFieldName } from 'SuiteResultTypes';
import { FocusSelectors } from 'focused';

/**
 * Checks if context has included tests
 */
export function useHasOnliedTests(
  testObject: TIsolateTest,
  fieldName?: TFieldName
): boolean {
  return isNotNullish(
    Walker.findClosest(testObject, (child: TIsolate) => {
      if (!FocusSelectors.isIsolateFocused(child)) return false;

      return FocusSelectors.isOnlyFocused(child, fieldName);
    })
  );
}
