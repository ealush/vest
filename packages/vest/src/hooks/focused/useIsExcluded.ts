import {
  Nullable,
  isNotEmpty,
  isNotNullish,
  optionalFunctionValue,
} from 'vest-utils';
import { Isolate, Walker } from 'vestjs-runtime';

import { FocusModes } from 'FocusedKeys';
import { IsolateTest } from 'IsolateTest';
import { useInclusion } from 'SuiteContext';
import { IsolateFocused } from 'focused';
import { isIsolateFocused } from 'isIsolateFocused';
import { useIsExcludedIndividually } from 'skipWhen';
//Checks whether a certain test profile excluded by any of the exclusion groups.

function useClosestMatchingFocus(
  testObject: IsolateTest
): Nullable<IsolateFocused> {
  return Walker.findClosest(testObject, (child: Isolate) => {
    if (!isIsolateFocused(child)) return false;

    return child.match?.includes(testObject.fieldName);
  });
}

function hasFocus(focus: Nullable<IsolateFocused>) {
  return isNotEmpty(focus?.match);
}

function isSkipFocused(focus: Nullable<IsolateFocused>): boolean {
  return focus?.focusMode === FocusModes.SKIP && hasFocus(focus);
}

function isOnlyFocused(focus: Nullable<IsolateFocused>): boolean {
  return (
    focus?.focusMode === FocusModes.ONLY &&
    (hasFocus(focus) || focus?.matchAll === true)
  );
}

// eslint-disable-next-line complexity, max-statements
export function useIsExcluded(testObject: IsolateTest): boolean {
  const { fieldName } = testObject;

  if (useIsExcludedIndividually()) return true;

  const inclusion = useInclusion();

  const focusMatch = useClosestMatchingFocus(testObject);

  // if test is skipped
  // no need to proceed
  if (isSkipFocused(focusMatch)) return true;

  const isTestIncluded = isOnlyFocused(focusMatch);

  // if field is only'ed
  if (isTestIncluded) return false;

  // If there is _ANY_ `only`ed test (and we already know this one isn't) return true
  if (useHasIncludedTests(testObject)) {
    // Check if inclusion rules for this field (`include` hook)
    return !optionalFunctionValue(inclusion[fieldName]);
  }

  // We're done here. This field is not excluded
  return false;
}

/**
 * Checks if context has included tests
 */
function useHasIncludedTests(testObject: IsolateTest): boolean {
  return isNotNullish(
    Walker.findClosest(testObject, (child: Isolate) => {
      if (!isIsolateFocused(child)) return false;

      return isOnlyFocused(child);
    })
  );
}
