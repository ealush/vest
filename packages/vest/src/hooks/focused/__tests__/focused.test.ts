import { faker } from '@faker-js/faker';
import { ErrorStrings } from 'ErrorStrings';

import { dummyTest } from '../../../../testUtils/testDummy';

import { skip, only } from 'focused';
import { group } from 'group';
import { TTestSuite } from 'testUtils/TVestMock';
import { useIsExcluded } from 'useIsExcluded';
import * as vest from 'vest';

let res: boolean, res1: boolean;

enum FieldNames {
  f1 = 'f1',
  f2 = 'f2',
  f3 = 'f3',
}

describe('focused hooks', () => {
  test('isExcluded should respect group exclusion', () => {
    let testObject: vest.IsolateTest;
    let testObject1: vest.IsolateTest;

    const validate = vest.create(() => {
      testObject = dummyTest.failing();

      group('group_1', () => {
        vest.skip(true);
        testObject1 = dummyTest.failing();
      });

      res = useIsExcluded(testObject);
      res1 = useIsExcluded(testObject1);
    });

    validate();

    expect(res).toBe(false);
    expect(res1).toBe(true);
  });

  describe('`only` hook', () => {
    describe('string input', () => {
      test('isExcluded returns false for included field', () => {
        vest.create(() => {
          vest.only(FieldNames.f1);
          const test = dummyTest.failing(FieldNames.f1);
          res = useIsExcluded(test);
        })();
        expect(res).toBe(false);
      });

      test('isGroupExcluded returns false for included groups', () => {
        vest.create(() => {
          vest.only('group_name');

          group('group_name', () => {
            res = useIsGroupExcluded('group_name');
          });
        })();
        expect(res).toBe(false);
      });

      test('isExcluded returns true for non included field', () => {
        vest.create(() => {
          vest.only(FieldNames.f1);
          const t1 = dummyTest.failing(FieldNames.f1);
          const t2 = dummyTest.failing(FieldNames.f2);
          res1 = useIsExcluded(t1);
          res = useIsExcluded(t2);
        })();
        expect(res1).toBe(false);
        expect(res).toBe(true);
      });

      test('isGroupExcluded returns true for non included group', () => {
        vest.create(() => {
          vest.only.group('group_1');

          group('group_1', jest.fn());
          group('group_2', jest.fn());

          res1 = useIsGroupExcluded('group_1');
          res = useIsGroupExcluded('group_2');
        })();

        expect(res1).toBe(false);
        expect(res).toBe(true);
      });
    });

    describe('array input', () => {
      test('isExcluded returns false for included field', () => {
        vest.create(() => {
          vest.only([FieldNames.f1, FieldNames.f2]);
          const test1 = dummyTest.failing(FieldNames.f1);
          const test2 = dummyTest.failing(FieldNames.f2);
          res = useIsExcluded(test1);
          res1 = useIsExcluded(test2);
        })();
        expect(res).toBe(false);
        expect(res1).toBe(false);
      });

      test('isGroupExcluded returns false for included groups', () => {
        const results: boolean[] = [];
        vest.create(() => {
          vest.only.group(['group_1', 'group_2']);

          group('group_1', jest.fn());
          group('group_2', jest.fn());
          group('group_3', jest.fn());

          results.push(
            useIsGroupExcluded('group_1'),
            useIsGroupExcluded('group_2'),
            useIsGroupExcluded('group_3')
          );
        })();
        expect(results).toEqual([false, false, true]);
      });

      test('isExcluded returns true for non included field', () => {
        const results: boolean[] = [];
        const results1: boolean[] = [];
        vest.create(() => {
          const test1 = dummyTest.failing(FieldNames.f1);
          const test2 = dummyTest.failing(FieldNames.f2);
          const test3 = dummyTest.failing(FieldNames.f3);
          results.push(
            useIsExcluded(test1),
            useIsExcluded(test2),
            useIsExcluded(test3)
          );
          vest.only([FieldNames.f1, test2.fieldName]);
          results1.push(
            useIsExcluded(test1),
            useIsExcluded(test2),
            useIsExcluded(test3)
          );
        })();
        expect(results).toEqual([false, false, false]);
        expect(results1).toEqual([false, false, true]);
      });

      test('isGroupExcluded returns true for non included groups', () => {
        const results: boolean[] = [];
        vest.create(() => {
          vest.only.group(['group_1', 'group_2']);

          group('group_3', jest.fn());
          results.push(
            useIsGroupExcluded('group_1'),
            useIsGroupExcluded('group_2'),
            useIsGroupExcluded('group_3')
          );
        })();
        expect(results).toEqual([false, false, true]);
      });
    });
  });

  describe('`skip` hook', () => {
    describe('string input', () => {
      test('isExcluded returns true for excluded field', () => {
        vest.create(() => {
          vest.skip(FieldNames.f1);
          const test1 = dummyTest.failing(FieldNames.f1);
          res = useIsExcluded(test1);
        })();
        expect(res).toBe(true);
      });

      test('isGroupExcluded returns true for excluded groups', () => {
        vest.create(() => {
          vest.skip.group('group_1');
          res = useIsGroupExcluded('group_1');
          res1 = useIsGroupExcluded('group_2');
        })();

        expect(res).toBe(true);
        expect(res1).toBe(false);
      });

      test('isExcluded returns false for non excluded field', () => {
        vest.create(() => {
          vest.skip(FieldNames.f1);
          res = useIsExcluded(vest.test(FieldNames.f2, jest.fn()));
        })();
        expect(res).toBe(false);
      });
    });

    test('isGroupExcluded returns false for tests in non excluded groups', () => {
      vest.create(() => {
        vest.skip('group_1');
        res = useIsExcluded(vest.test('field_1', jest.fn()));
      })();
      expect(res).toBe(false);
    });

    describe('array input', () => {
      test('isExcluded returns true for excluded field', () => {
        vest.create(() => {
          vest.skip([FieldNames.f1, FieldNames.f2]);
          const test1 = dummyTest.failing(FieldNames.f1);
          const test2 = dummyTest.failing(FieldNames.f2);
          res = useIsExcluded(test1);
          res1 = useIsExcluded(test2);
        })();
        expect(res).toBe(true);
        expect(res1).toBe(true);
      });

      test('isGroupExcluded returns true for excluded groups', () => {
        const results: boolean[] = [];
        vest.create(() => {
          vest.skip.group(['group_1', 'group_2']);
          results.push(
            useIsGroupExcluded('group_1'),
            useIsGroupExcluded('group_2')
          );
        })();
        expect(results).toEqual([true, true]);
      });

      test('isExcluded returns false for non included field', () => {
        vest.create(() => {
          vest.skip([FieldNames.f1, FieldNames.f2]);
          const test3 = dummyTest.failing(FieldNames.f3);
          res = useIsExcluded(test3);
        })();
        expect(res).toBe(false);
      });

      test('isGroupExcluded returns false for non excluded groups', () => {
        vest.create(() => {
          vest.skip(['group_1', 'group_2']);
          res = useIsGroupExcluded('group_3');
        })();
        expect(res).toBe(false);
      });
    });

    describe('Field is in a non included group', () => {
      let suite: TTestSuite;

      beforeEach(() => {
        suite = vest.create(() => {
          vest.only.group('group_1');

          vest.group('group_1', () => {
            vest.test('field_1', jest.fn());
          });
          vest.group('group_2', () => {
            vest.test('field_2', jest.fn());
          });
        });
        suite();
      });

      it('Should exclude test', () => {
        expect(suite.get().tests.field_2.testCount).toBe(0);
      });
    });
  });

  describe('Error handling', () => {
    describe('When called outside of a suite', () => {
      it('Should throw an error', () => {
        expect(() => only(faker.random.word())).toThrow(
          ErrorStrings.HOOK_CALLED_OUTSIDE
        );
        expect(() => skip(faker.random.word())).toThrow(
          ErrorStrings.HOOK_CALLED_OUTSIDE
        );
      });
    });
  });
});

// FIXME: These are lower quality tests, and we need to replace them with more robust ones
// describe('isExcluded', () => {
//   let exclusion: Partial<TExclusion>;

//   const runIsExcluded = (
//     exclusion: Partial<TExclusion>,
//     testObject: IsolateTest
//   ) => {
//     vest.staticSuite(() => {

//     })
//     // SuiteContext.run({}, () => {
//     //   Object.assign(useExclusion(), exclusion);
//     //   const res = useIsExcluded(testObject);

//     //   return res;
//     // });
//   }

//   const genTest = (fieldName: string, groupName?: string) =>
//     new IsolateTest({
//       fieldName,
//       testFn: jest.fn(),
//       groupName,
//     });
//   describe('skip', () => {
//     beforeEach(() => {
//       exclusion = { tests: { field_1: false, field_2: false } };
//     });
//     it('returns true for skipped field', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_1'))).toBe(
//         true
//       );
//     });
//     it('returns false for non skipped field', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(false);
//       expect(runIsExcluded(exclusion, genTest('field_4', 'group_1'))).toBe(
//         false
//       );
//     });
//   });
//   describe('only', () => {
//     beforeEach(() => {
//       exclusion = { tests: { field_1: true, field_2: true } };
//     });
//     it('returns false for included field', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(false);
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_1'))).toBe(
//         false
//       );
//     });
//     it('returns true for non included field', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_4', 'group_1'))).toBe(
//         true
//       );
//     });
//   });
//   describe('only+skip', () => {
//     beforeEach(() => {
//       exclusion = {
//         tests: { field_1: true, field_2: true, field_3: false, field_4: false },
//       };
//     });
//     it('returns false for included tests', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(false);
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_1'))).toBe(
//         false
//       );
//     });
//     it('returns true excluded tests', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_4', 'group_1'))).toBe(
//         true
//       );
//     });
//     it('returns true for non included field', () => {
//       expect(runIsExcluded(exclusion, genTest('field_5'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_6', 'group_1'))).toBe(
//         true
//       );
//     });
//   });
//   describe('skip.group', () => {
//     beforeEach(() => {
//       exclusion = {
//         groups: {
//           group_1: false,
//           group_2: false,
//         },
//       };
//     });

//     it('Returns true for tests in skipped group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_2'))).toBe(
//         true
//       );
//     });
//     it('Returns false for tests in non skipped groups', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3', 'group_3'))).toBe(
//         false
//       );
//       expect(runIsExcluded(exclusion, genTest('field_4', 'group_4'))).toBe(
//         false
//       );
//     });
//     it('Returns false for tests outside of any group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(false);
//       expect(runIsExcluded(exclusion, genTest('field_4'))).toBe(false);
//     });
//   });
//   describe('only.group', () => {
//     beforeEach(() => {
//       exclusion = {
//         groups: {
//           group_1: true,
//           group_2: true,
//         },
//       };
//     });

//     it('returns false for tests in included groups', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
//         false
//       );
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_2'))).toBe(
//         false
//       );
//     });

//     it('returns true for groups in non included groups', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_3'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_4'))).toBe(
//         true
//       );
//     });

//     it('returns true for tests outside of any group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_2'))).toBe(true);
//     });
//   });

//   describe('only.group + only', () => {
//     beforeEach(() => {
//       exclusion = {
//         groups: {
//           group_1: true,
//           group_2: true,
//         },
//         tests: { field_1: true, field_2: true },
//       };
//     });

//     it('returns true for included tests outside of the group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_2'))).toBe(true);
//     });

//     it('returns false for included tests in included groups', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
//         false
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_2'))).toBe(
//         false
//       );
//     });

//     it('returns true for included test in non included group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_3'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_4'))).toBe(
//         true
//       );
//     });

//     it('returns true for non included test in included group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3', 'group_1'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_4', 'group_2'))).toBe(
//         true
//       );
//     });

//     it('returns true for non included tests', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_4'))).toBe(true);
//     });
//   });

//   describe('skip.group + only', () => {
//     beforeEach(() => {
//       exclusion = {
//         groups: {
//           group_1: false,
//           group_2: false,
//         },
//         tests: { field_1: true, field_2: true },
//       };
//     });

//     it('returns true for included tests in excluded groups', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_1'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_3', 'group_1'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_2'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_2'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_3', 'group_2'))).toBe(
//         true
//       );
//     });

//     it('returns false for included tests in non excluded groups', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_3'))).toBe(
//         false
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_3'))).toBe(
//         false
//       );
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_4'))).toBe(
//         false
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_4'))).toBe(
//         false
//       );
//     });

//     it('returns true for non included tests', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_4'))).toBe(true);
//     });

//     it('returns false for included tests outside of the group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(false);
//       expect(runIsExcluded(exclusion, genTest('field_2'))).toBe(false);
//     });
//   });

//   describe('only.group + skip', () => {
//     beforeEach(() => {
//       exclusion = {
//         groups: {
//           group_1: true,
//           group_2: true,
//         },
//         tests: { field_1: false, field_2: false },
//       };
//     });

//     it('returns true for excluded tests', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(true);
//       expect(runIsExcluded(exclusion, genTest('field_2'))).toBe(true);
//     });
//     it('returns true for excluded test in included group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_2'))).toBe(
//         true
//       );
//     });
//     it('returns true for excluded test in non included group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_1', 'group_3'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_2', 'group_4'))).toBe(
//         true
//       );
//     });
//     it('returns false for non excluded test in included group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3', 'group_1'))).toBe(
//         false
//       );
//       expect(runIsExcluded(exclusion, genTest('field_4', 'group_2'))).toBe(
//         false
//       );
//     });
//     it('returns true for non excluded test in non included group', () => {
//       expect(runIsExcluded(exclusion, genTest('field_3', 'group_3'))).toBe(
//         true
//       );
//       expect(runIsExcluded(exclusion, genTest('field_4', 'group_4'))).toBe(
//         true
//       );
//     });
//   });

//   describe('Invalid input', () => {
//     describe('ExclusionItem is not a string', () => {
//       it('Should skip registration', () => {
//         const suite = vest.create('suite', () => {
//           // @ts-ignore - Invalid input
//           vest.only(111);

//           vest.test('f1', () => false);
//         });

//         expect(suite().hasErrors('f1')).toBe(true);
//       });
//     });
//   });
// });
