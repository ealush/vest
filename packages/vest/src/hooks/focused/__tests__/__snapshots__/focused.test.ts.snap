// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Level Focus Top Level Skip When passing false Should run all fields 1`] = `
{
  "done": [Function],
  "dump": [Function],
  "errorCount": 3,
  "errors": [
    SummaryFailure {
      "fieldName": "F1",
      "groupName": undefined,
      "message": "F1 error",
    },
    SummaryFailure {
      "fieldName": "F2",
      "groupName": undefined,
      "message": "F2 error",
    },
    SummaryFailure {
      "fieldName": "F3",
      "groupName": undefined,
      "message": "F3 error",
    },
  ],
  "getError": [Function],
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarning": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isPending": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "pendingCount": 0,
  "suiteName": undefined,
  "testCount": 3,
  "tests": {
    "F1": SummaryBase {
      "errorCount": 1,
      "errors": [
        "F1 error",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
    "F2": SummaryBase {
      "errorCount": 1,
      "errors": [
        "F2 error",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
    "F3": SummaryBase {
      "errorCount": 1,
      "errors": [
        "F3 error",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
  },
  "valid": false,
  "warnCount": 0,
  "warnings": [],
}
`;

exports[`Top Level Focus Top Level Skip When passing undefined Should run all fields 1`] = `
{
  "done": [Function],
  "dump": [Function],
  "errorCount": 3,
  "errors": [
    SummaryFailure {
      "fieldName": "F1",
      "groupName": undefined,
      "message": "F1 error",
    },
    SummaryFailure {
      "fieldName": "F2",
      "groupName": undefined,
      "message": "F2 error",
    },
    SummaryFailure {
      "fieldName": "F3",
      "groupName": undefined,
      "message": "F3 error",
    },
  ],
  "getError": [Function],
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarning": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isPending": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "pendingCount": 0,
  "suiteName": undefined,
  "testCount": 3,
  "tests": {
    "F1": SummaryBase {
      "errorCount": 1,
      "errors": [
        "F1 error",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
    "F2": SummaryBase {
      "errorCount": 1,
      "errors": [
        "F2 error",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
    "F3": SummaryBase {
      "errorCount": 1,
      "errors": [
        "F3 error",
      ],
      "pendingCount": 0,
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": [],
    },
  },
  "valid": false,
  "warnCount": 0,
  "warnings": [],
}
`;
